Pass on the GPU acceleration section of Chapter 13 (#505)
browserengineering/book: book/animations.md

A GPU is essentially a chip in your computer that runs programs, much
like your CPU. But the GPU is specialized toward running very simple
programs with massive parallelism; it was developed to apply simple
operations, in parallel, for every pixel on the screen. This makes GPUs faster for
drawing simple shapes and *much* faster for applying visual effects.

At a high level, to raster and draw on the GPU our browser
must:[^gpu-variations]

[^gpu-variations]: These steps vary a bit in the details by GPU architecture.

* *Upload* the display list to specialized GPU memory.

* *Compile* GPU programs that raster and draw the display list.[^compiled-gpu]

[^compiled-gpu]: That's right, GPU programs are dynamically compiled! This
allows GPU programs to be portable across a wide variety of implementations
that may have very different instruction sets or acceleration tactics.
Of course, between frames these compiled programs will typically be
cached, so this step won't occur on every frame.

* *Raster* every drawing command into GPU textures.[^texture]

[^texture]: A surface represented on the GPU is called a *texture*. There can be

> more than one surface, and practically speaking they often can't be rastered in

Reviewer: ```suggestion
more than one texture, and practically speaking they often can't be rastered in
```

parallel with each other.

* *Draw* the textures onto the screen.

SDL and Skia support all of these steps---in fact, it's mostly a
matter of passing them the right parameters. Let's do that. Note that
a real browser typically implements both CPU and GPU raster and draw,
because in some cases CPU raster and draw can be faster than using the

> GPU.[^example-cpu-fast] In our browswer, for simplicity, we'll stick

Reviewer: ```suggestion
GPU.[^example-cpu-fast] In our browser, for simplicity, we'll stick
```

to GPU mode for all pages.

[^example-cpu-fast]: Any of the four steps can make GPU raster and

>  draw slow. Large display list take a while to upload. Complex

Reviewer: ```suggestion
 draw slow. Large display lists take a while to upload. Complex
```

GPU acceleration
================

If you try the fade animation in our browser, you'll probably notice
that it's not particularly smooth. And that shouldn't be surprising;
after all, [last chapter](scheduling.md#profiling-rendering) had the
following figures for our browser:

    Time in raster-and-draw on average: 66ms
    Time in render on average: 23ms

With 66ms per frame, our browser is barely doing fifteen frames per
second, not sixty. For smooth animations we'll need to speed up raster
and draw.

The best way to do that is to move raster and draw to the [GPU][gpu].
A GPU is essentially a chip in your computer that runs programs, much
like your CPU. But the GPU is specialized toward running very simple

> programs with massive parallelism; it was developed to do simple

Reviewer: ```suggestion
programs with massive parallelism; it was developed to apply simple
```

> operations to every pixel on the screen. This makes GPUs faster for

Reviewer: ```suggestion
operations, in parallel, for every pixel on the screen. This makes GPUs faster for
```

* *Compile* GPU programs that raster and draw the display list.[^compiled-gpu]

[^compiled-gpu]: That's right, GPU programs are dynamically compiled! This
allows GPU programs to be portable across a wide variety of implementations
that may have very different instruction sets or acceleration tactics.
Of course, between frames these compiled programs will typically be
cached, so this step won't occur on every frame.

* *Raster* every drawing command into GPU textures.[^texture]

[^texture]: A surface represented on the GPU is called a *texture*. There can be
more than one surface, and practically speaking they often can't be rastered in
parallel with each other.

* *Draw* the textures onto the screen.

SDL and Skia support all of these steps---in fact, it's mostly a
matter of passing them the right parameters. Let's do that. Note that
a real browser typically implements both CPU and GPU raster and draw,
because in some cases CPU raster and draw can be faster than using the
GPU.[^example-cpu-fast] In our browswer, for simplicity, we'll stick
to GPU mode for all pages.

[^example-cpu-fast]: For example, the upload step isn't necessary on a

>     CPU, nor is the upload step. And GPUs typically have a lot less

Reviewer: ```suggestion
    CPU, nor is the compile step. And GPUs typically have a lot less
```

[^compiled-gpu]: That's right, GPU programs are dynamically compiled! This
allows GPU programs to be portable across a wide variety of implementations
that may have very different instruction sets or acceleration tactics.

* For each desired surface, *execute* the raster into GPU textures.[^texture]

[^texture]: A surface represented on the GPU is called a *texture*. There can be
more than one surface, and practically speaking they often can't be rastered in
parallel with each other.

* *Draw* the textures onto the screen.

You can configure any GPU program to draw directly to the *screen framebuffer*,
or to an *intermediate texture*.[^gpu-texture] Draw proceeds bottom-up: the
leaf textures are generated from raw display list data structures, and internal
surface tree textures are generated by computing visual effects on one or more
children. The root of the tree is the framebuffer texture.

[^gpu-texture]: Recall from [Chapter 11](visual-effects.md) that there can be
surfaces that draw into other surfaces, forming a tree. Skia internally does
this sometimes, based on various triggers such as blend modes. The internal
nodes draw into intermediate textures.

> The time to run GPU raster is roughly the sum of the time for these four steps.

Reviewer: I think this paragraph is good to keep, as it explains how slowdowns can come from each part. I suppose we don't actually use this knowledge though. Footnote?

Author: I merged this text into the `example-cpu-fast` footnote.

        sdl2.SDL_GL_SwapWindow(self.sdl_window)
```

> Finally, in our browser also creates a Skia surfaces for the

Reviewer: ```suggestion
Finally, our browser also creates a Skia surfaces for the
```

        sdl2.SDL_GL_SwapWindow(self.sdl_window)
```

Finally, our browser also creates a Skia surfaces for the
`chrome_surface` and `tab_surface`. We don't want to draw these
straight to the screen, so the incantation is a bit different:

``` {.python}
class Browser:
    def __init__(self):
        # ...
        self.chrome_surface =  skia.Surface.MakeRenderTarget(
                self.skia_context, skia.Budgeted.kNo,
                skia.ImageInfo.MakeN32Premul(WIDTH, CHROME_PX))
        assert self.chrome_surface is not None
```

Again, you should think of these changes mostly as boilerplate, since
the details of GPU operation aren't our focus here.[^color-space] Make
sure to apply the same treatment to `tab_surface` (with different
width and height arguments).

[^color-space]: For example, a different color space that is required
for GPU mode.

Thanks to SDL's and Skia's thorough support for GPU rendering, that
should be all that's necessary for our browser to raster and draw on
the GPU. And as expected, speed is much improved:

    Time in raster-and-draw on average: 24ms
    Time in render on average: 23ms

> That's almost three times faster, almost fast enough to hit sixty

Reviewer: ```suggestion
That's almost three times faster, almost fast enough to hit sixty
```
```suggestion
That's almost three times faster, almost fast enough to hit sixty
```
```suggestion
That's about three times faster, almost fast enough to hit sixty
```

frames per second. And on your computer, you'll likely see even more
speedup than that, so perhaps your animations are already quite
smooth. But if we want to go faster yet, we'll need to reduce the

> total amount of raster and draw work we're doing.

Reviewer: ```suggestion
total amount of work in raster and draw.
```
